crates/storage/provider/src/post_state.rs

```diff
@@ -338,66 +272,94 @@ impl PostState {
+            .retain(|block_number, _| *block_number > revert_to_block);
 
         non_reverted_state
     }
 
     /// Add a newly created account to the post-state.
-    pub fn create_account(&mut self, address: Address, account: Account) {
-        self.add_and_apply(Change::AccountCreated {
-            id: self.current_transition_id,
-            address,
-            account,
-        });
+    pub fn create_account(
+        &mut self,
+        block_number: BlockNumber,
+        address: Address,
+        account: Account,
+    ) {
+        self.accounts.insert(address, Some(account));
+        self.account_changes.entry(block_number).or_default().entry(address).or_insert(None);
     }
 
     /// Add a changed account to the post-state.
     ///
     /// If the account also has changed storage values, [PostState::change_storage] should also be
     /// called.
-    pub fn change_account(&mut self, address: Address, old: Account, new: Account) {
-        self.add_and_apply(Change::AccountChanged {
-            id: self.current_transition_id,
-            address,
-            old,
-            new,
-        });
+    pub fn change_account(
+        &mut self,
+        block_number: BlockNumber,
+        address: Address,
+        old: Account,
+        new: Account,
+    ) {
+        self.accounts.insert(address, Some(new));
+        self.account_changes.entry(block_number).or_default().entry(address).or_insert(Some(old));
     }
 
     /// Mark an account as destroyed.
-    pub fn destroy_account(&mut self, address: Address, account: Account) {
-        self.add_and_apply(Change::AccountDestroyed {
-            id: self.current_transition_id,
-            address,
-            old: account,
-        });
-        self.add_and_apply(Change::StorageWiped { id: self.current_transition_id, address });
+    pub fn destroy_account(
+        &mut self,
+        block_number: BlockNumber,
+        address: Address,
+        account: Account,
+    ) {
+        self.accounts.insert(address, None);
+        self.account_changes
+            .entry(block_number)
+            .or_default()
+            .entry(address)
+            .or_insert(Some(account));
+        let storage = self.storage.entry(address).or_default();
+        storage.wiped = true;
+        storage.storage.clear();
+        let storage_changes =
+            self.storage_changes.entry(block_number).or_default().entry(address).or_default();
+        storage_changes.wiped = true;
     }
 
     /// Add changed storage values to the post-state.
-    pub fn change_storage(&mut self, address: Address, changeset: StorageChangeset) {
-        self.add_and_apply(Change::StorageChanged {
-            id: self.current_transition_id,
-            address,
-            changeset,
-        });
+    pub fn change_storage(
+        &mut self,
+        block_number: BlockNumber,
+        address: Address,
+        changeset: StorageChangeset,
+    ) {
+        self.storage
```

Comment:
nit - there seems to be a slight repetition here, maybe it makes sense to have a helper like before with an enum?

Reply:
I would leave this as it is as to not introduce conflict, we can probably take another refactor pass on `PostState` in the future.